File: Java/SPARK/Limits/src/main/java/frc/robot/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.SparkBase.PersistMode;
import com.revrobotics.spark.SparkBase.ResetMode;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.spark.SparkLimitSwitch;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.SparkMaxConfig;
import com.revrobotics.spark.config.LimitSwitchConfig.Type;
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;

import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

public class Robot extends TimedRobot {
  private SparkMax motor;
  private SparkMaxConfig motorConfig;
  private SparkLimitSwitch forwardLimitSwitch;
  private SparkLimitSwitch reverseLimitSwitch;
  private RelativeEncoder encoder;

  public Robot() {
    /*
     * Initialize the SPARK MAX and get its limit switch and encoder objects for
     * later use.
     */
    motor = new SparkMax(1, MotorType.kBrushless);
    forwardLimitSwitch = motor.getForwardLimitSwitch();
    reverseLimitSwitch = motor.getReverseLimitSwitch();
    encoder = motor.getEncoder();

    /*
     * Create a new SPARK MAX configuration object. This will store the
     * configuration parameters for the SPARK MAX that we will set below.
     */
    motorConfig = new SparkMaxConfig();

    // Set the idle mode to brake to stop immediately when reaching a limit
    motorConfig.idleMode(IdleMode.kBrake);

    // Enable limit switches to stop the motor when they are closed
    motorConfig.limitSwitch
        .forwardLimitSwitchType(Type.kNormallyOpen)
        .forwardLimitSwitchEnabled(true)
        .reverseLimitSwitchType(Type.kNormallyOpen)
        .reverseLimitSwitchEnabled(true);

    // Set the soft limits to stop the motor at -50 and 50 rotations
    motorConfig.softLimit
        .forwardSoftLimit(50)
        .forwardSoftLimitEnabled(true)
        .reverseSoftLimit(-50)
        .reverseSoftLimitEnabled(true);

    /*
     * Apply the configuration to the SPARK MAX.
     *
     * kResetSafeParameters is used to get the SPARK MAX to a known state. This
     * is useful in case the SPARK MAX is replaced.
     *
     * kPersistParameters is used to ensure the configuration is not lost when
     * the SPARK MAX loses power. This is useful for power cycles that may occur
     * mid-operation.
     */
    motor.configure(motorConfig, ResetMode.kResetSafeParameters, PersistMode.kNoPersistParameters);

    // Reset the position to 0 to start within the range of the soft limits
    encoder.setPosition(0);

    // Initialize dashboard values
    SmartDashboard.setDefaultBoolean("Direction", true);
  }

  @Override
  public void robotPeriodic() {
    // Display data from SPARK onto the dashboard
    SmartDashboard.putBoolean("Forward Limit Reached", forwardLimitSwitch.isPressed());
    SmartDashboard.putBoolean("Reverse Limit Reached", reverseLimitSwitch.isPressed());
    SmartDashboard.putNumber("Applied Output", motor.getAppliedOutput());
    SmartDashboard.putNumber("Position", encoder.getPosition());
  }

  @Override
  public void autonomousInit() {
  }

  @Override
  public void autonomousPeriodic() {
  }

  @Override
  public void teleopInit() {
  }

  @Override
  public void teleopPeriodic() {
    // Set the motor setpoint based on the direction from the dashboard
    if (SmartDashboard.getBoolean("Direction", true)) {
      motor.set(0.2);
    } else {
      motor.set(-0.2);
    }
  }

  @Override
  public void disabledInit() {
  }

  @Override
  public void disabledPeriodic() {
  }

  @Override
  public void testInit() {
  }

  @Override
  public void testPeriodic() {
  }

  @Override
  public void simulationInit() {
  }

  @Override
  public void simulationPeriodic() {
  }
}

File: C++/SPARK/Limits/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Robot.h"

#include <frc/smartdashboard/SmartDashboard.h>
#include <rev/config/SparkMaxConfig.h>

Robot::Robot() {
  /*
   * Create a new SPARK MAX configuration object. This will store the
   * configuration parameters for the SPARK MAX that we will set below.
   */
  SparkMaxConfig motorConfig;

  // Set the idle mode to brake to stop immediately when reaching a limit
  motorConfig.SetIdleMode(SparkMaxConfig::IdleMode::kBrake);

  // Enable limit switches to stop the motor when they are closed
  motorConfig.limitSwitch
      .ForwardLimitSwitchType(LimitSwitchConfig::Type::kNormallyOpen)
      .ForwardLimitSwitchEnabled(true)
      .ReverseLimitSwitchType(LimitSwitchConfig::Type::kNormallyOpen)
      .ReverseLimitSwitchEnabled(true);

  // Set the soft limits to stop the motor at -50 and 50 rotations
  motorConfig.softLimit.ForwardSoftLimit(50)
      .ForwardSoftLimitEnabled(true)
      .ReverseSoftLimit(-50)
      .ReverseSoftLimitEnabled(true);

  /*
   * Apply the configuration to the SPARK MAX.
   *
   * kResetSafeParameters is used to get the SPARK MAX to a known state. This
   * is useful in case the SPARK MAX is replaced.
   *
   * kPersistParameters is used to ensure the configuration is not lost when
   * the SPARK MAX loses power. This is useful for power cycles that may occur
   * mid-operation.
   */
  m_motor.Configure(motorConfig, SparkMax::ResetMode::kResetSafeParameters,
                    SparkMax::PersistMode::kNoPersistParameters);

  // Reset the position to 0 to start within the range of the soft limits
  m_encoder.SetPosition(0);

  // Initialize dashboard values
  frc::SmartDashboard::SetDefaultBoolean("Direction", true);
}

void Robot::RobotPeriodic() {
  // Display data from SPARK onto the dashboard
  frc::SmartDashboard::PutBoolean("Forward Limit Reached",
                                  m_forwardLimitSwitch.Get());
  frc::SmartDashboard::PutBoolean("Reverse Limit Reached",
                                  m_reverseLimitSwitch.Get());
  frc::SmartDashboard::PutNumber("Applied Output", m_motor.GetAppliedOutput());
  frc::SmartDashboard::PutNumber("Position", m_encoder.GetPosition());
}

void Robot::AutonomousInit() {}
void Robot::AutonomousPeriodic() {}

void Robot::TeleopInit() {}
void Robot::TeleopPeriodic() {
  // Set the motor setpoint based on the direction from the dashboard
  if (frc::SmartDashboard::GetBoolean("Direction", true)) {
    m_motor.Set(0.2);
  } else {
    m_motor.Set(-0.2);
  }
}

void Robot::DisabledInit() {}
void Robot::DisabledPeriodic() {}

void Robot::TestInit() {}
void Robot::TestPeriodic() {}

void Robot::SimulationInit() {}
void Robot::SimulationPeriodic() {}

#ifndef RUNNING_FRC_TESTS
int main() { return frc::StartRobot<Robot>(); }
#endif
